apps/dashboard/src/components/workflow-editor/steps/configure-step-template-form.tsx

```diff
@@ -27,55 +28,65 @@ export type StepEditorProps = {
   step: StepDataDto;
 };
 
-export type ConfigureStepTemplateFormProps = StepEditorProps & {
-  debouncedUpdate: (data: UpdateWorkflowDto) => void;
+type ConfigureStepTemplateFormProps = StepEditorProps & {
+  update: (data: UpdateWorkflowDto) => void;
 };
-export const ConfigureStepTemplateForm = (props: ConfigureStepTemplateFormProps) => {
-  const { workflow, step, debouncedUpdate } = props;
 
+export const ConfigureStepTemplateForm = (props: ConfigureStepTemplateFormProps) => {
+  const { workflow, step, update } = props;
   const schema = useMemo(() => buildDynamicZodSchema(step.controls.dataSchema ?? {}), [step.controls.dataSchema]);
-
   const defaultValues = useMemo(
     () => merge(buildDefaultValues(step.controls.uiSchema ?? {}), step.controls.values),
-    []
+    [step]
   );
 
   const form = useForm({
-    mode: 'onChange',
     resolver: zodResolver(schema),
     defaultValues: defaultValues,
-    shouldFocusError: true,
+    shouldFocusError: false,
+  });
+
+  const { onBlur, flushFormUpdates } = useFormAutosave({
+    previousData: defaultValues,
+    form,
+    save: (data) => {
+      update({
+        ...workflow,
+        steps: workflow.steps.map((s) => {
```

Comment:
We will have to tweak steps and other nested structures several times. I suggest creating helper functions to handle such data changes. For example `update(buildNewStep(...))`;

This is not a blocker for this PR.